src/v/cluster/members_backend.cc

```diff
@@ -325,11 +386,28 @@ void members_backend::calculate_reallocations_after_node_added(
                   && to_move.left_to_move > 0) {
                     partition_reallocation reallocation(
                       model::ntp(tp_ns.ns, tp_ns.tp, p.id), p.replicas.size());
-
                     reallocation.replicas_to_remove.emplace(to_move.id);
+                    auto current_assignment = p;
+                    reassign_replicas(current_assignment, reallocation);
+                    // if this reassignment does not involve the node we are
+                    // targetting do not add it
+                    if (!is_reassigned_to_node(reallocation, meta.update.id)) {
+                        continue;
+                    }
+                    reallocation.state = reallocation_state::reassigned;
                     meta.partition_reallocations.push_back(
                       std::move(reallocation));
                     to_move.left_to_move--;
+                    // reached max concurrent reallocations, yield
+                    if (
+                      meta.partition_reallocations.size()
+                      >= _max_concurrent_reallocations()) {
+                        vlog(
+                          clusterlog.info,
+                          "reached limit of max concurrent reallocations: {}",
+                          meta.partition_reallocations.size());
+                        co_return;
```

Comment:
What is the stop condition for finishing the process of node addition? I can't quite understand where we decide if we should schedule another batch, or just finish.

Reply:
there are two conditions: 
1) no allocations possible moving partitions to the new node
2) all nodes having even number of partitions